// fs-filesystem/esm/utilities.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
// --------------------------------------
// Common Core - Utility Functions
// --------------------------------------

var lasti = exports.lasti = function lasti(a) {
  return a[a.length - 1];
};
var stringify = exports.stringify = function stringify(v) {
  return v && v.toString ? v.toString() : '' + v;
};
var hasSubstr = exports.hasSubstr = function hasSubstr(s, sub) {
  return s.indexOf(sub) !== -1;
};
var getYesNo = exports.getYesNo = function getYesNo(v) {
  return v === 'Yes' ? true : v === 'No' ? false : undefined;
};

// --------------------------------------
// Common Core - Empty FS Objects
// --------------------------------------

var emptyDevice = exports.emptyDevice = function emptyDevice() {
  return {
    id: null,
    node: null,
    whole: false,
    parent: null,
    name: null,
    size: null,
    description: null,
    protocol: null,
    blockSize: null,
    readOnly: null,
    removable: null
  };
};

var emptyVolume = exports.emptyVolume = function emptyVolume() {
  return {
    id: null,
    node: null,
    whole: false,
    parent: null,
    name: null,
    description: null,
    blockSize: null,
    blocks: null,
    readOnly: null,
    mounted: null,
    mountPoint: null,
    partitionType: null,
    fs: null,
    space: { total: null, available: null, used: null }
  };
};
